src/Language/Futhark/Unused.hs

```diff
@@ -0,0 +1,207 @@
+locsInExp (ArrayLit exs _ _) = M.unions $ map locsInExp exs
+locsInExp (Attr _ ex _) = locsInExp ex
+locsInExp (Project _ ex _ _) = locsInExp ex
+locsInExp (Not ex _) = locsInExp ex
+locsInExp (Assert ex1 ex2 _ _) = locsInExp ex2 `M.union` locsInExp ex1
+locsInExp (Constr _ exs _ _) = M.unions $ map locsInExp exs
+locsInExp (Update ex1 _ ex2 _) = locsInExp ex2 `M.union` locsInExp ex1
+locsInExp (RecordUpdate ex1 _ ex2 _ _) = locsInExp ex2 `M.union` locsInExp ex1
+locsInExp (Lambda _ ex _ _ _) = locsInExp ex
+-- locsInExp (OpSection (QualName _ vn) _ _)  = M.adjust (S.insert vn)
+locsInExp (OpSectionLeft _ _ ex _ _ _) = locsInExp ex
+locsInExp (OpSectionRight _ _ ex _ _ _) = locsInExp ex
+locsInExp (Ascript ex _ _) = locsInExp ex
+-- locsInExp (Var (QualName _ vn) _ _)  = M.adjust (S.insert vn)
+locsInExp (AppExp app _) =
+  case app of
+    Apply ex1 lst _ ->
+      M.unions (NE.toList $ NE.map (locsInExp . snd) lst) `M.union` locsInExp ex1
+    Coerce ex _ _ -> locsInExp ex
+    Range ex1 mb_ex inc_ex _ -> M.unions $ map locsInExp (ex1 : fromInc inc_ex : maybeToList mb_ex)
+    LetPat _ _ ex1 ex2 _ -> locsInExp ex2 `M.union` locsInExp ex1
+    LetFun vn (_, _, _, _, ex1) ex2 loc -> M.insert vn loc $ locsInExp ex2 `M.union` locsInExp ex1 -- Important case! function defn
+    If ex1 ex2 ex3 _ -> locsInExp ex3 `M.union` locsInExp ex2 `M.union` locsInExp ex1
+    DoLoop _ _ ex1 loop_ex ex2 _ -> locsInExp (fromLoop loop_ex) `M.union` locsInExp ex2 `M.union` locsInExp ex1
+    BinOp _ _ (ex1, _) (ex2, _) _ -> locsInExp ex2 `M.union` locsInExp ex1
+    LetWith _ _ sl ex1 ex2 _ -> M.unions $ map locsInExp $ fromSlice sl <> [ex1, ex2]
+    Index ex sl _ -> M.unions $ map locsInExp $ ex : fromSlice sl
+    Match ex cases _ -> M.unions $ map locsInExp $ ex : map fromCase (NE.toList cases)
+locsInExp _ = M.empty
+
+locsInModExp :: ModExpBase Info VName -> LocMap
+locsInModExp (ModParens mex _) = locsInModExp mex
+locsInModExp (ModDecs dbs _) = M.unions $ map locsInDec dbs
+locsInModExp (ModApply mex1 mex2 _ _ _) = locsInModExp mex1 `M.union` locsInModExp mex2
+locsInModExp (ModAscript mex _ _ _) = locsInModExp mex
+locsInModExp (ModLambda _ _ mex _) = locsInModExp mex
+locsInModExp _ = M.empty
+
+fromInc :: Inclusiveness (ExpBase Info VName) -> ExpBase Info VName
+fromInc (DownToExclusive x) = x
+fromInc (ToInclusive x) = x
+fromInc (UpToExclusive x) = x
+
+fromLoop :: LoopFormBase f vn -> ExpBase f vn
+fromLoop (For _ x) = x
+fromLoop (ForIn _ x) = x
+fromLoop (While x) = x
+
+fromSlice :: [DimIndexBase f vn] -> [ExpBase f vn]
+fromSlice = concatMap fromDimInd
+
+fromDimInd :: DimIndexBase f vn -> [ExpBase f vn]
+fromDimInd (DimFix x) = [x]
+fromDimInd (DimSlice m1 m2 m3) = catMaybes [m1, m2, m3]
+
+fromCase :: CaseBase f vn -> ExpBase f vn
+fromCase (CasePat _ x _) = x
+
+getFieldExp :: FieldBase Info VName -> ExpBase Info VName
+getFieldExp (RecordFieldExplicit _ ex _) = ex
+getFieldExp _ = error "placeholder" -- TODO: provide an appropriate error here.
```

Comment:
What's going on here?

Reply:
it's used when reading record literals. to avoid code duplication. It's used on L78 and L139 current commit

Reply:
Using `error` is an absolute last resort.  Duplicate code is better.

Reply:
I just realized it was an obvious use for `Maybe`. Fixed.